crypt.go

```diff
@@ -237,6 +238,132 @@ func getPassword(sourceDevicePath, description string, reader io.Reader) (string
+// authorized to access the data on the decrypted volume with the device
+// mapper name returned by VolumeName.
+func (c *SnapModelChecker) IsModelAuthorized(model SnapModel) (bool, error) {
+	return c.keyData.IsSnapModelAuthorized(c.auxKey, model)
+}
+
+// VolumeName is the device mapper name of the volume associated with this
+// SnapModelChecker.
+func (c *SnapModelChecker) VolumeName() string {
+	return c.volumeName
+}
+
+type activateWithKeyDataError struct {
+	k   *KeyData
+	err error
+}
+
+func (e *activateWithKeyDataError) Error() string {
+	return fmt.Sprintf("%s: %v", e.k.ID(), e.err)
+}
+
+func (e *activateWithKeyDataError) Unwrap() error {
+	return e.err
+}
+
+type keyDataAndError struct {
+	*KeyData
+	err error
+}
+
+type activateWithKeyDataState struct {
+	volumeName       string
+	sourceDevicePath string
+	activateOptions  []string
+	keyringPrefix    string
+
+	keys []*keyDataAndError
+
+	keyData *KeyData
+	auxKey  AuxiliaryKey
+}
+
+func (s *activateWithKeyDataState) errors() (out []*activateWithKeyDataError) {
+	for _, k := range s.keys {
+		if k.err == nil {
+			continue
+		}
+		out = append(out, &activateWithKeyDataError{k: k.KeyData, err: k.err})
+	}
+	return
+}
+
+func (s *activateWithKeyDataState) snapModelChecker() *SnapModelChecker {
+	return &SnapModelChecker{s.volumeName, s.keyData, s.auxKey}
+}
+
+func (s *activateWithKeyDataState) tryActivateWithRecoveredKey(keyData *KeyData, key DiskUnlockKey, auxKey AuxiliaryKey) error {
+	if err := activate(s.volumeName, s.sourceDevicePath, key, s.activateOptions); err != nil {
+		return xerrors.Errorf("cannot activate volume: %w", err)
+	}
+
+	s.keyData = keyData
+	s.auxKey = auxKey
+
+	if err := keyring.AddKeyToUserKeyring(key, s.sourceDevicePath, keyringPurposeDiskUnlock, s.keyringPrefix); err != nil {
+		fmt.Fprintf(os.Stderr, "secboot: Cannot add key to user keyring: %v\n", err)
+	}
+
+	if err := keyring.AddKeyToUserKeyring(auxKey, s.sourceDevicePath, keyringPurposeAuxiliary, s.keyringPrefix); err != nil {
+		fmt.Fprintf(os.Stderr, "secboot: Cannot add key to user keyring: %v\n", err)
+	}
+
+	return nil
+}
+
+func (s *activateWithKeyDataState) tryKeyDataAuthModeNone(k *KeyData) error {
+	key, auxKey, err := k.RecoverKeys()
+	if err != nil {
+		return xerrors.Errorf("cannot recover key: %w", err)
+	}
+
+	return s.tryActivateWithRecoveredKey(k, key, auxKey)
+}
+
+func (s *activateWithKeyDataState) run() (success bool) {
+	// Try keys that don't require any additional authentication first
+	for _, k := range s.keys {
+		if k.AuthMode() != AuthModeNone {
+			continue
+		}
+
+		if err := s.tryKeyDataAuthModeNone(k.KeyData); err != nil {
+			k.err = err
+			continue
+		}
+
+		return true
+	}
+
+	// TODO: Passphrase support
+
+	// We've failed at this point
+	return false
+}
+
+func makeActivateWithKeyDataState(volumeName, sourceDevicePath string, activateOptions []string, keyringPrefix string, keys []*KeyData) *activateWithKeyDataState {
```

Comment:
this would be more commonly called new.... as it has also behavior attached

Reply:
I've renamed this.